desc/compute/utils.py

```diff
@@ -683,63 +813,124 @@ def surface_integrals(grid, q=jnp.array([1]), surface_label="rho", max_surface=F
+            a scalar function over the previously mentioned domain.
+
+            When ``q`` is 2-dimensional, the intention is to integrate,
+            over the domain parameterized by rho, theta, and zeta,
+            a vector-valued function over the previously mentioned domain.
+
+            When ``q`` is 3-dimensional, the intention is to integrate,
+            over the domain parameterized by rho, theta, and zeta,
+            a matrix-valued function over the previously mentioned domain.
+
+        Returns
+        -------
+        integrals : ndarray
+            Surface integral of the input over each surface in the grid.
+
+        """
+        integrands = (spacing * jnp.nan_to_num(q).T).T
+        # `integrands` may have shape (g.size, f.size, v.size), where
+        #     g is the grid function depending on the integration variables
+        #     f is a function which may be independent of the integration variables
+        #     v is the vector of components of f (or g).
+        # The intention is to integrate `integrands` which is a
+        #     vector-valued            (with v.size components)
+        #     function-valued          (with image size of f.size)
+        #     function over the grid   (with domain size of g.size = grid.num_nodes)
+        # over each surface in the grid.
+
+        # The distinction between f and v is semantic.
+        # We may alternatively consider an `integrands` of shape (g.size, f.size) to
+        # represent a vector-valued (with f.size components) function over the grid.
+        # Likewise, we may alternatively consider an `integrands` of shape
+        # (g.size, v.size) to represent a function-valued (with image size v.size)
+        # function over the grid. When `integrands` has dimension one, it is a
+        # scalar function over the grid. That is, a
+        #     vector-valued            (with 1 component),
+        #     function-valued          (with image size of 1)
+        #     function over the grid   (with domain size of g.size = grid.num_nodes)
+
+        # The integration is performed by applying `masks`, the surface
+        # integral operator, to `integrands`. This operator hits the matrix formed
+        # by the last two dimensions of `integrands`, for every element along the
+        # previous dimension of `integrands`. Therefore, when `integrands` has three
+        # dimensions, the second must hold g. We may choose which of the first and
+        # third dimensions hold f and v. The choice below transposes `integrands` to
+        # shape (v.size, g.size, f.size). As we expect f.size >> v.size, the
+        # integration is in theory faster since numpy optimizes large matrix
+        # products. However, timing results showed no difference.
+        axis_to_move = (integrands.ndim == 3) * 2
+        integrals = jnp.moveaxis(
+            masks @ jnp.moveaxis(integrands, axis_to_move, 0), 0, axis_to_move
```

Comment:
The GitHub CI failed (twice) a new augmented Lagrangian test because a value is positive with order $10^{-12}$ when it should be negative. I can't reproduce the failure locally with `pytest -k test_constrained_AL_scalar tests/test_optimizer.py`. I'll try testing on Princeton's clusters.

Seems unlikely but I'll try changing `@` to `jnp.matmul(precision=Precision.HIGH)` and see if it fixes.